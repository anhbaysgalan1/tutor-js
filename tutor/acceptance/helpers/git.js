const gitP = require('simple-git/promise');
const path = require('path');
const fs = require('fs-extra');
const { imagesDir } = require('./images-dir');
const chalk = require('chalk');

const repoURL = 'https://github.com/openstax/tutor-images.git';

function initialiseRepo(git) {
  console.log(chalk.yellow(`Cloning images repo to ${imagesDir}`));
  return git.clone(repoURL, imagesDir);
}

function getCurrentBranch() {
  const git = gitP(__dirname);
  return new Promise((resolve, reject) => {
    git.branchLocal((err, branch) => {
      if (err) { reject(err); }
      resolve(branch.current);
    });
  });
}

async function gitCommit() {
  const g = git();
  const currentBranchName = await getCurrentBranch();
  const branchInfo = await g.branch();
  if (branchInfo.all.includes(currentBranchName)) {
    await g.checkout(currentBranchName);
  } else {
    await g.checkoutLocalBranch(currentBranchName);
  }
  await g.add(['--all', '.']);
  await g.commit(`updated ${currentBranchName}`);
  await g.push('origin', currentBranchName);
}

async function gitInit() {
  await fs.ensureDir(imagesDir);
  const git = gitP(imagesDir);
  return git.cwd(imagesDir)
    .then(() => git.checkIsRepo())
    .then(isRepo => isRepo ? git : initialiseRepo(git))
    .then(() => git.fetch())
    .then(() => git);
}

function git() {
  process.exit(1)
  return gitP(imagesDir);
}

module.exports = { git, gitCommit, gitInit, getCurrentBranch };
