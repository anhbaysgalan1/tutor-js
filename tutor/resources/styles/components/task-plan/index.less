@import './guide';
@import './reading';
@import './builder';
@import './task-plan-stats';
@import './homework';
@import './reading';
@import './chapter-section';
@import './lms-info';
@import './mini-editor';
@import './teacher-task-plans-listing';
@import './controls';

.task-plan {
  .panel-default {
    border: none;
  }

  .panel-heading {
    border-bottom: none;
  }

  .select-topics {
    > .panel-body {
      overflow: auto;
      height:  ~"calc(100vh - 200px)";
    }
  }

  .tutor-plan-set(header, ' .dialog > .panel-heading');

  > .dialog {
    .tutor-pad-panel();
  }
}
