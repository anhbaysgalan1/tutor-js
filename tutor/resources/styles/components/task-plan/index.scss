@import './guide';
@import './reading';
@import './builder';
@import './task-plan-stats';
@import './homework';
@import './reading';
@import './chapter-section';
@import './lms-info';
@import './mini-editor';
@import './teacher-task-plans-listing';
@import './controls';


.task-plan {
  .panel-default {
    border: none;
  }

  .panel-heading {
    border-bottom: none;
    height: 50px;
    display: flex;
    align-items: center;
    justify-content: space-between;
    button.close {
      top: 0;
      transistion: opacity 0.3s;
    }
  }

  .select-topics {
    > .panel-body {
      overflow: auto;
      max-height: calc(100vh - 200px);
    }
  }

  @include tutor-plan-set(header, '.dialog > .panel-heading');

  > .dialog {
    @include tutor-pad-panel();
  }
}
