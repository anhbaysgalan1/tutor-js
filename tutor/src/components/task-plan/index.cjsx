React = require 'react'
{HomeworkPlan} = require './homework'
{ReadingPlan} = require './reading'
{ExternalPlan} = require './external'
{EventPlan} = require './event'
LoadableItem = require '../loadable-item'
{CourseStore, CourseActions} = require '../../flux/course'
{TaskPlanStore, TaskPlanActions} = require '../../flux/task-plan'
Router = require '../../helpers/router'
{ default: TourRegion } = require '../tours/region'

{ScrollToMixin} = require 'shared'

PLAN_TYPES =
  reading: ReadingPlan
  homework: HomeworkPlan
  external: ExternalPlan
  event: EventPlan

getPlanType = (typeName) ->
  type = PLAN_TYPES[typeName]

HomeworkShell = React.createClass
  displayName: 'HomeworkShell'
  contextTypes:
    router: React.PropTypes.object

  render: ->
    {courseId, id} = Router.currentParams()
    <PlanShell courseId={courseId} id={id} type='homework'/>

ReadingShell = React.createClass
  displayName: 'ReadingShell'
  contextTypes:
    router: React.PropTypes.object

  render: ->
    {courseId, id} = Router.currentParams()
    <PlanShell courseId={courseId} id={id} type='reading'/>

ExternalShell = React.createClass
  displayName: 'ExternalShell'
  contextTypes:
    router: React.PropTypes.object

  render: ->
    {courseId, id} = Router.currentParams()
    <PlanShell courseId={courseId} id={id} type='external'/>

EventShell = React.createClass
  displayName: 'EventShell'
  contextTypes:
    router: React.PropTypes.object

  render: ->
    {courseId, id} = Router.currentParams()
    <PlanShell courseId={courseId} id={id} type='event'/>

PlanBuilder = ({ id, courseId, body: Body, type }) ->
  <TourRegion
    id={"#{type}-assignment-editor"}
    courseId={courseId}
  >
    <Body id={id} courseId={courseId} />
  </TourRegion>


PlanShell = React.createClass
  displayName: 'PlanShell'

  mixins: [ScrollToMixin]

  contextTypes:
    router: React.PropTypes.object

  componentDidMount:  -> @scrollToTop()

  getInitialState: ->
    {id} = Router.currentParams()
    type = @props.type
    if not getPlanType(type)
      @context.router.transitionTo(Router.makePathname('NotFoundRoute'))
      return

    if not id or id is 'new'
      id = TaskPlanStore.freshLocalId()
      TaskPlanActions.create(id, {type})

    {id}

  getId: -> @state.id or Router.currentParams().id

  getType: ->
    typeName = @props.type
    getPlanType(typeName)

  render: ->
    {courseId} = Router.currentParams()
    id = @getId()

    if TaskPlanStore.isDeleteRequested(id)
      return <Type id={id} courseId={courseId} />

    <LoadableItem
      id={id}
      store={TaskPlanStore}
      actions={TaskPlanActions}
      renderItem={=>
        <PlanBuilder type={@props.type} body={@getType()} id={id} courseId={courseId} />
      }
    />

module.exports = {ReadingShell, HomeworkShell, ExternalShell, EventShell}
