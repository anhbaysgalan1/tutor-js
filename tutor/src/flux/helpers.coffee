_ = require 'underscore'
flux = require 'flux-react'

LOADING = 'loading'
LOADED  = 'loaded'
FAILED  = 'failed'
SAVING  = 'saving'
DELETING = 'deleting'
DELETED = 'deleted'

STATES =
  LOADING : LOADING
  LOADED  : LOADED
  FAILED  : FAILED
  SAVING  : SAVING
  DELETING: DELETING
  DELETED : DELETED

idCounter = 0
CREATE_KEY = -> "_CREATING_#{idCounter++}"

isNew = (id) -> /_CREATING_/.test(id)

CrudConfig = ->
  {
    _asyncStatus: {}
    _local: {}
    _changed: {}
    _errors: {}
    _reload: {}

    # If the specific type needs to do something else to the object:
    # _loaded : (obj, id) ->
    # _saved : (obj, id) ->
    # _reset : ->

    reset: ->
      @_asyncStatus = {}
      @_local = {}
      @_changed = {}
      @_errors = {}
      @_reload = {}

      @_reset?()
      @emitChange()

    dontReload: (id) ->
      @_asyncStatus[id] is LOADED and @_HACK_DO_NOT_RELOAD

    FAILED: (status, msg, id, args...) ->
      @_asyncStatus[id] = FAILED
      @_errors[id] = msg
      @_failed?({status, msg}, id, args...)

      unless status is 0 # indicates network failure
        delete @_local[id]
        @emitChange()

    load: (id) ->
      @_load(id)

    # allow aliasing load without triggering api call.
    _load: (id) ->
      # Add a shortcut for unit testing
      return if @dontReload(id)
      @_reload[id] = false
      @_asyncStatus[id] = LOADING
      @emitChange()

    loaded: (obj, id, args...) ->
      # id = obj.id
      @_asyncStatus[id] = LOADED

      if obj
        # If the specific type needs to do something else to the object:
        @_local[id] = @_loaded?(obj, id, args...) or obj

      @emitChange()

    save: (id, obj) ->
      # Note: id could be isNew()
      @_asyncStatus[id] = SAVING
      @emitChange()

    saved: (result, id, args...) ->
      # id = result.id
      @_asyncStatus[id] = LOADED # TODO: Maybe make this SAVED

      # If the specific type needs to do something else to the object:
      obj = @_saved?(result, id, args...)
      result = obj if obj

      if result
        @_local[id] = result
        @_local[result.id] = result
        delete @_changed[result.id]
      else
        console.warn('API WARN: Server did not return JSON after saving. Patching locally')
        # Merge all the local changes into the new local object
        @_local[id] = _.extend(@_local[id], @_changed[id])
      delete @_changed[id]
      delete @_errors[id]
      # If the specific type needs to do something else to the object:
      @emitChange()

    create: (localId, attributes = {}) ->
      throw new Error('BUG: MUST provide a local id') unless isNew(localId)
      @_local[localId] = {}
      @_changed[localId] = attributes
      @_asyncStatus[localId] = LOADED

    created: (result, localId, args...) ->
      @_local[localId] = result # HACK: So react component can still manipulate the same object
      @_local[result.id] = result
      @_asyncStatus[localId] = LOADED
      @_asyncStatus[result.id] = LOADED
      @emitChange()
      @_created?(result, localId, args...)

    _change: (id, obj) ->
      @_changed[id] ?= {}
      _.extend(@_changed[id], obj)
      @emitChange()

    _save: (id) ->
      @_asyncStatus[id] = SAVING

    delete: (id) ->
      @_asyncStatus[id] = DELETING
      @_delete?(id)

    deleted: (result, id, args...) ->
      @_asyncStatus[id] = DELETED
      delete @_local[id]
      @emitChange()
      @_deleted?(result, id, args...)

    clearChanged: (id) ->
      delete @_changed[id]

    HACK_DO_NOT_RELOAD: (bool) -> @_HACK_DO_NOT_RELOAD = bool

    # Keep this here so other exports method have access to it
    _get: (id) ->
      val = @_local[id]
      return null unless val or @_asyncStatus[id] is SAVING
      # Scores stores an Array unlike most other stores
      if _.isArray(val)
        val
      else
        _.extend({}, val, @_changed[id])

    exports:
      isUnknown: (id) -> not @_asyncStatus[id]
      isLoading: (id) -> @_asyncStatus[id] is LOADING
      isLoaded: (id) -> @_asyncStatus[id] is LOADED
      isDeleting: (id) -> @_asyncStatus[id] is DELETING
      isSaving: (id) -> @_asyncStatus[id] is SAVING
      isFailed: (id) -> @_asyncStatus[id] is FAILED
      getAsyncStatus: (id) -> @_asyncStatus[id]
      get: (id) -> @_get(id)
      isChanged: (id) -> not _.isEmpty(@_changed[id])
      getChanged: (id) -> @_changed[id] or {}
      freshLocalId: -> CREATE_KEY()
      isNew: (id) -> isNew(id)
      reload: (id) -> @_reload[id]
      getError: (id) -> @_errors[id]

  }

# Helper for creating a simple store for actions
makeSimpleStore = (storeConfig) ->

  actionsConfig = _.omit storeConfig, (value, key) ->
    typeof value isnt 'function' or key is 'exports'

  actionsConfig = _.keys(actionsConfig)
  actions = flux.createActions(actionsConfig)
  storeConfig.actions = _.values(actions)
  store = flux.createStore(storeConfig)
  {actions, store}


extendConfig = (newConfig, origConfig) ->
  newConfig.exports ?= {}
  _.defaults(newConfig, origConfig)
  _.defaults(newConfig.exports, origConfig.exports)
  newConfig


makeStandardStore = (name, storeConfig) ->
  extendConfig(storeConfig, new CrudConfig())
  {actions, store} = makeSimpleStore(storeConfig)
  {"#{name}Actions":actions, "#{name}Store":store}


module.exports = {CrudConfig, makeSimpleStore, extendConfig, makeStandardStore, isNew, STATES}
