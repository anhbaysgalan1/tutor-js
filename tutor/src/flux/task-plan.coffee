# coffeelint: disable=no_empty_functions
_ = require 'underscore'
cloneDeep = require 'lodash/cloneDeep'
each = require 'lodash/each'
map = require 'lodash/map'
pick = require 'lodash/pick'
isEmpty = require 'lodash/isEmpty'
negate = require 'lodash/negate'
validator = require 'validator'
moment = require 'moment'
{CrudConfig, makeSimpleStore, extendConfig} = require './helpers'
{TocStore} = require './toc'
TimeHelper  = require '../helpers/time'
{ExerciseStore} = require './exercise'
{ default: Publishing } = require '../models/jobs/task-plan-publish'

{default: Courses} = require '../models/course'
ContentHelpers = require '../helpers/content'

planCrudConfig = new CrudConfig()

hasValue = negate(isEmpty)

TUTOR_SELECTIONS =
  default: 3
  max: 4
  min: 2

PLAN_TYPES =
  HOMEWORK: 'homework'
  READING: 'reading'
  EXTERNAL: 'external'
  EVENT: 'event'

DEFAULT_TYPE = PLAN_TYPES.READING

sortTopics = (topics) ->
  _.sortBy(topics, (topicId) ->
    topic = TocStore.getSectionInfo(topicId)
    ContentHelpers.chapterSectionToNumber(topic.chapter_section)
  )

BASE_PLANS =
  homework:
    is_feedback_immediate: false
    settings:
      page_ids: []
      exercise_ids: []
      exercises_count_dynamic: TUTOR_SELECTIONS.default
  reading:
    settings:
      page_ids: []
  external:
    settings:
      external_url: ''
  event:
    settings: {}

newTaskPlan = (attributes = {}) ->
  attributes.type ?= DEFAULT_TYPE

  if BASE_PLANS[attributes.type]?
    _.extend({}, attributes, cloneDeep(BASE_PLANS[attributes.type]))
  else
    {}

validateSettings = (taskPlan = {}) ->
  taskPlan.type ?= DEFAULT_TYPE

  expectedSettings = _.keys(BASE_PLANS[taskPlan.type].settings)
  taskPlan.settings = _.pick(taskPlan.settings, expectedSettings)

  taskPlan

TaskPlanConfig =

  _stats: {}
  _asyncStatusStats: {}
  _server_copy: {}
  _silentError: {}

  _loaded: (obj, planId) ->
    @_server_copy[planId] = JSON.stringify(obj) if _.isObject(obj)
    @emit("loaded.#{planId}")
    obj

  # Somewhere, the local copy gets taken apart and rebuilt.
  # Keep a copy of what was served.
  _getOriginal: (planId) ->
    if _.isString(@_server_copy[planId])
      JSON.parse(@_server_copy[planId])
    else
      {}

  create: (localId, attributes = {}) ->
    attributes = newTaskPlan(attributes)
    planCrudConfig.create.call(@, localId, attributes)

  _getPlan: (planId) ->
    obj = @_get(planId)
    obj = validateSettings(obj)
    obj

  FAILED: -> # used by API

  _delete: (id) ->
    @emit('deleting', id)

  _deleted: (result, id, args...) ->
    @emit('deleted', id, result)

  # Returns copies of the given property names from settings
  # Copies are returned so that the store can be reset
  _getClonedSettings: (id, names...) ->
    plan = @_getPlan(id)
    settings = {}
    for name in names
      settings[name] = _.clone(plan.settings[name])
    return settings

  _changeSettings: (id, attributes) ->
    plan = @_getPlan(id)
    @_change(id, settings: _.extend({}, plan.settings, attributes))

  replaceTaskings: (id, taskings) ->
    if taskings?
      @_change(id, {tasking_plans: taskings})
    else if @_changed[id]?.tasking_plans?
      delete @_changed[id].tasking_plans
      @emitChange()

  updateTutorSelection: (id, direction) ->
    {exercises_count_dynamic} = @_getClonedSettings(id, 'exercises_count_dynamic')

    exercises_count_dynamic += direction

    exercises_count_dynamic = Math.min(TUTOR_SELECTIONS.max, exercises_count_dynamic)
    exercises_count_dynamic = Math.max(TUTOR_SELECTIONS.min, exercises_count_dynamic)
    @_changeSettings(id, {exercises_count_dynamic})

  updateTitle: (id, title) ->
    @_change(id, {title})

  updateDescription:(id, description) ->
    plan = @_getPlan(id)
    @_change(id, {description: description})

  updateUrl: (id, external_url) ->
    @_change(id, {settings: {external_url}})

  setEvent: (id) ->
    @_change(id, {settings: {}})

  sortTopics: (id) ->
    {page_ids} = @_getClonedSettings(id, 'page_ids')
    @_changeSettings(id, page_ids: sortTopics(page_ids))

  addTopic: (id, topicId) ->
    {page_ids} = @_getClonedSettings(id, 'page_ids')
    page_ids.push(topicId) unless page_ids.indexOf(topicId) >= 0
    @_changeSettings(id, {page_ids})

  setImmediateFeedback: (id, is_feedback_immediate) ->
    @_change(id, {is_feedback_immediate})

  removeTopic: (id, topicId) ->
    {page_ids, type, exercise_ids} = @_getClonedSettings(id, 'page_ids', 'exercise_ids')
    index = page_ids?.indexOf(topicId)
    page_ids?.splice(index, 1)
    if (type is PLAN_TYPES.HOMEWORK)
      exercise_ids = ExerciseStore.removeTopicExercises(exercise_ids, topicId)
    @_changeSettings(id, {page_ids, exercise_ids })

  updateTopics: (id, page_ids) ->
    @_changeSettings(id, {page_ids})

  addExercise: (id, exercise) ->
    {exercise_ids} = @_getClonedSettings(id, 'exercise_ids')
    unless exercise_ids.indexOf(exercise.id) >= 0
      exercise_ids.push(exercise.id)
    @_changeSettings(id, {exercise_ids})
    @emit("change-exercise-#{exercise?.id}")

  removeExercise: (id, exercise) ->
    {exercise_ids} = @_getClonedSettings(id, 'exercise_ids')
    index = exercise_ids?.indexOf(exercise.id)
    exercise_ids?.splice(index, 1)
    @_changeSettings(id, {exercise_ids})
    @emit("change-exercise-#{exercise?.id}")

  updateExercises: (id, exercise_ids) ->
    @_changeSettings(id, {exercise_ids})

  moveReading: (id, topicId, step) ->
    {page_ids} = @_getClonedSettings(id, 'page_ids')
    curIndex = page_ids?.indexOf(topicId)
    newIndex = curIndex + step

    if (newIndex < 0)
      newIndex = 0
    if not (newIndex < page_ids.length)
      newIndex = page_ids.length - 1

    page_ids[curIndex] = page_ids[newIndex]
    page_ids[newIndex] = topicId

    @_changeSettings(id, {page_ids})

  moveExercise: (id, exercise, step) ->
    {exercise_ids} = @_getClonedSettings(id, 'exercise_ids')
    curIndex = exercise_ids?.indexOf(exercise.id)
    newIndex = curIndex + step

    if (newIndex < 0)
      newIndex = 0
    if not (newIndex < exercise_ids.length)
      newIndex = exercise_ids.length - 1

    exercise_ids[curIndex] = exercise_ids[newIndex]
    exercise_ids[newIndex] = exercise.id

    @_changeSettings(id, {exercise_ids})

  _getStats: (id) ->
    @_stats[id]

  loadStats: (id) ->
    delete @_stats[id]
    @_asyncStatusStats[id] = 'loading'
    @emitChange()

  loadedStats: (obj, id) ->
    @_stats[id] = obj
    @_asyncStatusStats[id] = 'loaded'
    @emitChange()

  publish: (id) ->
    @emit('publishing', id)
    @_change(id, {is_publish_requested: true})

  saveSilent: (id, obj) ->
    @_save(id)
    @emitChange()

  _saved: (obj, id) ->
    @emit("saved.#{id}", obj)


  erroredSilent: (obj, id, courseId) ->
    @emit('errored', obj)

  resetPlan: (id) ->
    @_local[id] = _.clone(@_getOriginal(id))
    @clearChanged(id)

  removeUnsavedDraftPlan: (id) ->
    return unless @exports.isNew.call(@, id)
    delete @_local[id]
    @clearChanged(id)
    @emitChange()

  _isDeleteRequested: (id) ->
    deleteStates = [
      'deleting'
      'deleted'
    ]
    deleteStates.indexOf(@_asyncStatus[id]) > -1


  createClonedPlan: (newPlanId, {planId, courseId, due_at}) ->
    original = @_local[planId]
    originalPlan = {}
    copyKeys = ['title', 'description', 'is_feedback_immediate', 'settings', 'ecosystem_id']
    each(copyKeys, (key) ->
      originalPlan[key] = cloneDeep(original[key])
      # need to prevent false from being returned and ending the each loop early.
      true
    )

    course = Courses.get(courseId)
    tasking_plans = course.periods.map((period) ->
      {
        opens_at: course.starts_at
        target_id: period.id,
        target_type: 'period',
        due_at: due_at
      }
    )

    plan = _.extend(newTaskPlan(type: original.type), originalPlan, {
      id: newPlanId
      cloned_from_id: planId
      tasking_plans: tasking_plans
    })

    @_local[newPlanId] = {}
    @_changed[newPlanId] = plan
    @emitChange()


  exports:
    hasTopic: (id, topicId) ->
      plan = @_getPlan(id)
      plan?.settings.page_ids?.indexOf(topicId) >= 0

    getTopics: (id) ->
      plan = @_getPlan(id)
      plan?.settings.page_ids

    isChangingToDueAt: (id) ->
      plan = @_changed[id]
      plan? and not plan.is_feedback_immediate

    isFeedbackImmediate: (id) ->
      plan = @_getPlan(id)
      plan?.is_feedback_immediate

    getEcosystemId: (id, courseId) ->
      plan = @_getPlan(id)
      plan.ecosystem_id or Courses.get(courseId)?.ecosystem_id

    hasExercise: (id, exerciseId) ->
      plan = @_getPlan(id)
      plan?.settings.exercise_ids?.indexOf(exerciseId) >= 0

    getExercises: (id) ->
      plan = @_getPlan(id)
      plan?.settings.exercise_ids

    exerciseCount: (id) ->
      plan = @_getPlan(id)
      plan?.settings.exercise_ids.length

    getDescription: (id) ->
      plan = @_getPlan(id)
      plan?.description

    isHomework: (id) ->
      plan = @_getPlan(id)
      plan.type is PLAN_TYPES.HOMEWORK

    isValid: (id) ->
      plan = @_getPlan(id)
      if (plan.type is 'reading')
        hasValue(plan.title) and hasValue(plan.settings) and hasValue(plan.settings.page_ids)
      else if (plan.type is 'homework')
        hasValue(plan.title) and hasValue(plan.settings) and hasValue(plan.settings.exercise_ids)
      else if (plan.type is 'external')
        hasValue(plan.title) and hasValue(plan.settings) and hasValue(plan.settings.external_url)
      else if (plan.type is 'event')
        hasValue(plan.title)
      else
        false

    isPublished: (id) ->
      plan = @_getPlan(id)
      !!plan?.is_published

    isDeleteRequested: (id) -> @_isDeleteRequested(id)

    isEditable: (id) ->
      # cannot be/being deleted
      not @_isDeleteRequested(id)

    isPublishing: (id) ->
      return false if @exports.isNew.call(@, id)
      @_changed[id]?.is_publishing or Publishing.isPublishing(@_getPlan(id))

    canDecreaseTutorExercises: (id) ->
      plan = @_getPlan(id)
      plan.settings.exercises_count_dynamic > TUTOR_SELECTIONS.min

    canIncreaseTutorExercises: (id) ->
      plan = @_getPlan(id)
      plan.settings.exercises_count_dynamic < TUTOR_SELECTIONS.max

    getTutorSelections: (id) ->
      plan = @_getPlan(id)
      plan.settings.exercises_count_dynamic

    getStats: (id) ->
      @_getStats(id)

    isStatsLoading: (id) -> @_asyncStatusStats[id] is 'loading'

    isStatsLoaded: (id) -> !! @_stats[id]

    isStatsFailed: (id) -> !! @_stats[id]

    hasChanged: (id) ->
      changed = @exports.getChanged.call(@, id)

      if @exports.isNew.call(@, id)
        # omit tasking plan changes if new plan, only check for changes in other fields
        changed = _.omit(changed, 'tasking_plans')
        defaultTaskPlan = newTaskPlan({type: changed.type})
        return not _.isEqual(changed, defaultTaskPlan)

      not _.isEmpty(changed)

extendConfig(TaskPlanConfig, planCrudConfig)
{actions, store} = makeSimpleStore(TaskPlanConfig)
module.exports = {TaskPlanActions:actions, TaskPlanStore:store}
