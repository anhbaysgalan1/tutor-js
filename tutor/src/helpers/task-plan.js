moment = require 'moment'
_ = require 'underscore'
{TaskPlanStore} = require '../flux/task-plan'

module.exports = {

  # returns the opens_at and due_at dates combined if they're identical,
  # separated by target_id if not
  dates: (plan, options = {}) ->
    return {} if _.isEmpty( plan?.tasking_plans )
    pickFn = if options.only
      (tp) -> _.pick(tp, options.only)
    else
      (tp) -> _.pick(tp, 'opens_at', 'due_at')

    first = _.first( plan.tasking_plans )
    isIdentical = _.every plan.tasking_plans, (tp) ->
      if options.only?
        tp[options.only] is first[options.only]
      else
        tp.opens_at is first.opens_at and tp.due_at is first.due_at

    if isIdentical
      all: pickFn(first)
    else
      _.reduce( plan.tasking_plans, (memo, tp) ->
        memo[tp.target_id] = pickFn(tp)
        memo
      , {})


  earliestDueDate: (plan) ->
    dates = _.map(plan?.tasking_plans, 'due_at')
    _.first(dates.sort()) or ''


  apiEndpointOptions: (id, courseId) ->
    task = TaskPlanStore.get(id)
    options = if TaskPlanStore.isNew(id)
      url: "courses/#{courseId}/plans"
      method: 'POST'
    else
      url: "plans/#{id}"
    if task?.ecosystem_id
      options.params = {ecosystem_id: task.ecosystem_id}
    options

}
