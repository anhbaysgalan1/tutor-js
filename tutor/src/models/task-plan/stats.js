import {
  BaseModel, belongsTo, identifiedBy, session, hasMany, field, identifier,
} from '../base';
import {
  get, flatMap, groupBy, find, isNil, isObject, isEmpty, keys,
} from 'lodash';
import { computed } from 'mobx';
import { lazyInitialize } from 'core-decorators';
import ChapterSection from '../chapter-section';

@identifiedBy('task-plan/stats/answer-stat')
class AnswerStat extends BaseModel {

  @session answer_id;
  @session selected_count;

  @belongsTo({ model: 'task-plan/stats/question' }) question;

  @computed get content() {
    return find(get(this.question.content, 'answers'), a => a.id == this.answer_id) || {};
  }

  @computed get id() { return this.answer_id; }

  @computed get correctness() {
    return this.content.correctness;
  }
  @computed get content_html() {
    return this.content.content_html;
  }
}

@identifiedBy('task-plan/stats/answer')
class Answer extends BaseModel {

  @session({ type: 'array' }) student_names;
  @session free_response;
  @session answer_id;

  @belongsTo({ model: 'task-plan/stats/question' }) question;
  @computed get selected_count() {
    return find(this.question.answer_stats, anst => anst.answer_id == this.answer_id).selected_count || 0;
  }

}


class ReviewQuestion {

  constructor(question) {
    this.q = question;
  }
  get id() { return this.q.question_id; }
  @computed get answers() { return this.q.answer_stats; }
  @computed get formats() { return this.q.content.formats; }
  @computed get stem_html(){ return this.q.content.stem_html; }
  @computed get stimulus_html() { return this.q.content.stimulus_html; }
}


@identifiedBy('task-plan/stats/question')
export class Question extends BaseModel {

  @session question_id;
  @session answered_count;

  @belongsTo({ model: 'task-plan/stats/exercise' }) exercise;
  @hasMany({ model: Answer, inverseOf: 'question' }) answers;
  @hasMany({ model: AnswerStat, inverseOf: 'question' }) answer_stats;
  @lazyInitialize forReview = new ReviewQuestion(this);

  @computed get hasFreeResponse() {
    return find(this.answers, a => !isEmpty(a.free_response));
  }

  @computed get content() {
    return find(this.exercise.contentData.questions, q =>
      q.id == this.question_id
    ) || {};
  }
}


@identifiedBy('task-plan/stats/exercise')
export class Exercise extends BaseModel {

  @session content;

  @session average_step_number;
  @belongsTo({ model: 'task-plan/stats/page' }) page;
  @hasMany({ model: Question, inverseOf: 'exercise' }) question_stats;

  @computed get contentData() {
    return (isNil(this.content) || isObject(this.content)) ? this.content : JSON.parse(this.content) || {};
  }

  @computed get uid() {
    return this.contentData ? this.contentData.uid : '';
  }
}


@identifiedBy('task-plan/stats/page')
export class Page extends BaseModel {

  @identifier id;
  @field({ model: ChapterSection }) chapter_section
  @session title;
  @session correct_count;
  @session incorrect_count;
  @session is_trouble;
  @session student_count;

  @hasMany({ model: Exercise, inverseOf: 'page' }) exercises;
}

@identifiedBy('task-plan/stats/stat')
export class Stats extends BaseModel {

  @session period_id;
  @session name;
  @session total_count;
  @session complete_count;
  @session partially_complete_count;
  @belongsTo({ model: 'task-plan/stats' }) taskPlan;
  @session is_trouble;
  @hasMany({ model: Page }) current_pages;
  @hasMany({ model: Page }) spaced_pages;

  @computed get exercises() {
    return flatMap(['current_pages', 'spaced_pages'], pageType => {
      return flatMap(this[pageType], pg => pg.exercises.peek());
    });
  }

  @computed get exercisesBySection() {
    return groupBy(this.exercises, ex => ex.page.chapter_section.asString);
  }

  @computed get sections() {
    return keys(this.exercisesBySection);
  }
}

@identifiedBy('task-plan/stats')
export default class TaskPlanStats extends BaseModel {

  @identifier id;
  @session title;
  @session type;

  @session shareable_url;

  @hasMany({ model: Stats, inverseOf: 'plan' }) stats;

  @belongsTo({ model: 'task-plan/teacher' }) taskPlan;

  fetch() { return { id: this.taskPlan.id }; }
  fetchReview() { return { id: this.taskPlan.id }; }

}
