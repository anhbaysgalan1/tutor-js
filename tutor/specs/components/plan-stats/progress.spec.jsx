import { React, SnapShot } from '../helpers/component-testing';
import Progress from '../../../src/components/plan-stats/progress';

import { Page } from '../../../src/models/task-plan/stats';

describe('TaskPlan stats progress bar', function() {
  let props;

  beforeEach(() => {
    props = {
      type: 'reading',
      activeSection: '1.1',
      data: new Page({
        id: 1,
        chapter_section: [1,2],
        correct_count: 1,
        incorrect_count: 1,
      }),
    };
  });

  it('matches snapshot', function() {
    const component = SnapShot.create(<Progress {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders as all incorrect', () => {
    props.data.correct_count = 0;
    const progress = mount(<Progress {...props} />);
    expect(progress).toHaveRendered('.reading-progress-bar[aria-valuenow=0]');
  });

});
