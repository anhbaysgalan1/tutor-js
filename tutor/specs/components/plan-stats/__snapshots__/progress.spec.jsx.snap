// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskPlan stats progress bar matches snapshot 1`] = `
<div
  className="reading-progress inactive"
>
  <div
    className="reading-progress-heading"
  >
    <strong>
      <span
        className="text-success"
      >
        1.2
      </span>
       
    </strong>
     
    <span
      className="reading-progress-student-count"
    >
      (
       students)
    </span>
  </div>
  <div
    className="reading-progress-container"
  >
    <div
      className="reading-progress-group progress"
    >
      <div
        alt="50% correct"
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={50}
        className="reading-progress-bar progress-bar-correct progress-bar"
        role="progressbar"
        style={
          Object {
            "width": "50%",
          }
        }
        type="correct"
      >
        50%
      </div>
      <div
        alt="50% incorrect"
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={50}
        className="reading-progress-bar progress-bar-incorrect progress-bar"
        role="progressbar"
        style={
          Object {
            "width": "50%",
          }
        }
        type="incorrect"
      >
        50%
      </div>
    </div>
  </div>
</div>
`;
